apps/hermes/server/src/state/aggregate.rs

```diff
@@ -568,6 +641,129 @@ where
+            .fetch_message_states(
+                vec![price_id.to_bytes()],
+                start_time.clone(),
+                MessageStateFilter::Only(MessageType::TwapMessage),
+            )
+            .await?;
+
+        let end_messages = state
+            .fetch_message_states(
+                vec![price_id.to_bytes()],
+                end_time.clone(),
+                MessageStateFilter::Only(MessageType::TwapMessage),
+            )
+            .await?;
+
+        if start_messages.is_empty() || end_messages.is_empty() {
+            continue;
+        }
+
+        let start_message = &start_messages[0];
+        let end_message = &end_messages[0];
+
+        if let (Message::TwapMessage(start_twap), Message::TwapMessage(end_twap)) =
+            (&start_message.message, &end_message.message)
+        {
+            match calculate_twap(start_twap, end_twap) {
+                Ok(twap_price) => {
+                    // down_slots_ratio describes the % of slots where the network was down
+                    // over the TWAP window. A value closer to zero indicates higher confidence.
+                    let total_slots = end_message.slot - start_message.slot;
+                    let total_down_slots = end_twap.num_down_slots - start_twap.num_down_slots;
+                    let down_slots_ratio = total_down_slots as f64 / total_slots as f64;
+
+                    // Add to calculated TWAPs
+                    calculated_twaps.push(PriceFeedTwap {
+                        id: *price_id,
+                        twap: twap_price,
+                        start_timestamp: start_twap.publish_time,
+                        end_timestamp: end_twap.publish_time,
+                        down_slots_ratio,
+                    });
+
+                    // Combine messages for update data
+                    let mut messages = Vec::new();
+                    messages.push(start_message.clone().into());
+                    messages.push(end_message.clone().into());
+
+                    if let Ok(update_data) = construct_update_data(messages) {
+                        all_update_data.push(update_data);
+                    } else {
+                        tracing::warn!(
+                            "Failed to construct update data for price feed {}",
+                            price_id
+                        );
+                        continue;
+                    }
+                }
+                Err(e) => {
+                    tracing::warn!(
+                        "Failed to calculate TWAP for price feed {}: {}",
+                        price_id,
+                        e
+                    );
+                    continue;
+                }
+            }
+        }
+    }
+
+    Ok(TwapsWithUpdateData {
+        twaps: calculated_twaps,
+        update_data: all_update_data,
+    })
+}
+fn calculate_twap(start_message: &TwapMessage, end_message: &TwapMessage) -> Result<Price> {
```

Comment:
I think there is another check that we need to perform here, and that's making sure that the given messages are the first one in their timestamp, and this can be checked by making sure `prev_publish_time != (or <) publish_time`

Reply:
Added the validation, but is it sufficient to simply throw an error in this case?